src/Symfony/Component/FeatureToggle/FeatureCollection.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\FeatureToggle;
+
+use Psr\Container\ContainerInterface;
+
+/** @implements \IteratorAggregate<int, Feature> */
+final class FeatureCollection implements ContainerInterface, \IteratorAggregate
+{
+    /** @var array<string, Feature>|null */
+    private array|null $features = null;
+
+    /** @var array<iterable<Feature>|(\Closure(): iterable<Feature>)> */
+    private array $featureProviders = [];
+
+    /**
+     * @param iterable<Feature> $features
+     */
+    public function __construct(iterable $features)
+    {
+        $this->append($features);
+    }
+
+    /**
+     * @param iterable<Feature>|(\Closure(): iterable<Feature>) $features
+     */
+    private function append(iterable|\Closure $features): void
+    {
+        $this->featureProviders[] = $features;
+        $this->features = null;
+    }
+
+    /**
+     * @param iterable<Feature>|(\Closure(): iterable<Feature>) $features
+     */
+    public function withFeatures(iterable|\Closure $features): self
+    {
+        $this->append($features);
+
+        return $this;
+    }
+
+    /**
+     * @phpstan-assert-if-true null $this->features
+     */
+    private function compile(): void
+    {
+        if (null !== $this->features) {
+            return;
+        }
+
+        $this->features = [];
+
+        foreach ($this->featureProviders as $featureProvider) {
+            if (is_callable($featureProvider)) {
+                $featureProvider = $featureProvider();
+            }
+
+            foreach ($featureProvider as $feature) {
+                $this->features[$feature->getName()] = $feature;
```

Comment:
Shouldn't we throw if 2 features have the same name?

Reply:
Maybe. I am not sure. Either we do some kind of last in "wins" or we ensure each is unique. But because we don't pull from teh providers during cache creation, we will figure this out only during runtime. I think it would be too much to throw an error because of this.

Reply:
To provide a good DX I would throw as proposed by @fancyweb

Reply:
Bear in mind that feature can be remotely provided. Thus making it impossible to know in advance what they will be without iterating over all providers. Because of this we tend to have to do this during runtime.

I think raising an exception during runtime would be bad DX. But we could make it a parameter `$throwOnError`. In the class it would default to false and in framework to `kernel.debug`.

WDYT ?

Reply:
The PR description does not show any explanation about how this remote loading is expected to work. As the public API of the component takes a feature name as input, modeling a concept of feature loader taking the feature name as argument might be a much better fit than a concept of provider returning an iterable forcing to load everything all the time.

Reply:
This would mean only one provider at a time ? Agree I can make things to not load everything all the time. Like stated in a previous comment :
- only iterate until I find a provider that provide the feature
- then instanciate the strategy classes only when calling the isEnabled (this can already be achieved by making the first strategy as lazy)